File: android/app/src/main/kotlin/com/nomadguy/noctra/lyricromanizer/LyricRomanizerTypes.kt
package com.nomadguy.noctra.lyricromanizer

/**
 * Port of lyric-romanizer/src/types.ts
 *
 * Core type definitions for the lyric romanizer engine.
 */

/** Supported script types for detection and romanization. */
enum class ScriptType {
    JAPANESE, CHINESE, KOREAN, CYRILLIC, DEVANAGARI, GUJARATI,
    TELUGU, KANNADA, ODIA, TAMIL, MALAYALAM, BENGALI, ARABIC, HEBREW,
    THAI, LATIN, OTHER
}

/** Chinese dialect option for romanization system selection. */
enum class ChineseDialect { MANDARIN, CANTONESE }

/** Options for romanization operations. */
data class RomanizeOptions(
    val script: ScriptType? = null,
    val dialect: ChineseDialect? = null
)

/** Result of a multi-line romanization operation. */
data class RomanizeResult(
    val script: ScriptType,
    val lines: List<String>,
    val fallbacks: List<Boolean>
)

/** Context passed to engine adapters during romanization. */
data class RomanizeEngineContext(
    val dialect: ChineseDialect = ChineseDialect.MANDARIN
)

/** An engine adapter: romanizes one line of its script. */
typealias RomanizeEngine = suspend (String, RomanizeEngineContext) -> String

/** Factory options for creating a Romanizer instance. */
data class RomanizerOptions(
    val japaneseDictPath: String? = null,
    val engines: Map<ScriptType, RomanizeEngine?>? = null
)

/** Thrown when attempting to romanize a script that has no engine. */
class UnsupportedRomanizationError(val script: ScriptType) :
    Exception("Script '${script.name}' requires external romanization.")

/** The main romanizer interface. */
interface Romanizer {
    suspend fun romanizeLine(line: String, options: RomanizeOptions? = null): String
    suspend fun romanizeLines(
        lines: List<String>,
        options: RomanizeOptions? = null
    ): RomanizeResult

    suspend fun warmup(scripts: List<ScriptType>? = null)
}

File: lib/services/lyric_romanizer/lyric_romanizer_types.dart
/// Port of lyric-romanizer/src/types.ts
///
/// Core type definitions for the lyric romanizer engine.

/// Supported script types for detection and romanization.
library;
enum ScriptType {
  japanese,
  chinese,
  korean,
  cyrillic,
  devanagari,
  gurmukhi,
  gujarati,
  telugu,
  kannada,
  odia,
  tamil,
  malayalam,
  bengali,
  arabic,
  hebrew,
  thai,
  latin,
  other,
}

/// Chinese dialect option for romanization system selection.
enum ChineseDialect { mandarin, cantonese }

/// Options for romanization operations.
class RomanizeOptions {
  final ScriptType? script;
  final ChineseDialect? dialect;
  const RomanizeOptions({this.script, this.dialect});
}

/// Result of a multi-line romanization operation.
class RomanizeResult {
  final ScriptType script;
  final List<String> lines;
  final List<bool> fallbacks;

  const RomanizeResult({
    required this.script,
    required this.lines,
    required this.fallbacks,
  });

  @override
  String toString() =>
      'RomanizeResult(script: ${script.name}, lines: $lines, fallbacks: $fallbacks)';
}

/// Context passed to engine adapters during romanization.
class RomanizeEngineContext {
  final ChineseDialect dialect;
  const RomanizeEngineContext({this.dialect = ChineseDialect.mandarin});
}

/// An engine adapter: romanizes one line of its script.
typedef RomanizeEngine = Future<String> Function(
  String line,
  RomanizeEngineContext context,
);

/// Factory options for creating a Romanizer instance.
class RomanizerOptions {
  final String? japaneseDictPath;
  final Map<ScriptType, RomanizeEngine?>? engines;
  const RomanizerOptions({this.japaneseDictPath, this.engines});
}

/// Thrown when attempting to romanize a script that has no engine.
class UnsupportedRomanizationError implements Exception {
  final ScriptType script;
  const UnsupportedRomanizationError(this.script);

  @override
  String toString() =>
      "UnsupportedRomanizationError: Script '${script.name}' requires external romanization.";
}

/// The main romanizer interface.
abstract class Romanizer {
  Future<String> romanizeLine(String line, {RomanizeOptions? options});
  Future<RomanizeResult> romanizeLines(
    List<String> lines, {
    RomanizeOptions? options,
  });
  Future<void> warmup([List<ScriptType>? scripts]);
}
